src/functions.php

```diff
@@ -377,13 +381,104 @@ function upload(string $xml, $config)
+ */
+function getOldPhonebook(Api $fritz, array $config)
+{
+    $formfields = [
+        'PhonebookId' => $config['phonebook']['id'],
+        'PhonebookExportName' => $config['phonebook']['name'],
+        'PhonebookExport' => "",
+    ];
+    $result = $fritz->postFile($formfields, []); // send the command to load existing phone book
+    if (substr($result, 0, 5) !== "<?xml") {
+        error_log("ERROR: Could not load old phonebook with ID=".$config['phonebook']['id']);
+        return new SimpleXMLElement("<xml><empty /></xml>");
+    }
+    $XMLPhonebook = simplexml_load_string($result);
+    return $XMLPhonebook;
+}
+
+
+/**
+ * Get quickdial and vanity special attributes from given XML phone book
+ *
+ * @param   SimpleXMLElement    $XMLPhonebook
+ * @return  array               [] or map with {phonenumber@CardDavUID}=>SimpleXMLElement-Attributes
+ */
+function getOldSpecialAttributes(SimpleXMLElement $XMLPhonebook)
+{
+    if (!property_exists($XMLPhonebook, "phonebook")) {
+        return [];
+    }
+
+    $specialAttributes = [];
+    foreach ($XMLPhonebook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
+            if ((isset($number->attributes()->quickdial) && $number->attributes()->quickdial >= 0)
+                || (isset($number->attributes()->vanity) && $number->attributes()->vanity != "")) {
+                // build unique key: {normalized-phone-number}@{vCard UUID} mapping to attributes (quickdial, vanity)
+                // normalizing means: remove all non-"+" and non-number characters like SPACE, MINUS, SLASH...
+                $key = preg_replace("/[^\+0-9]/", "", $number)."@".$contact->carddav_uid;
+                $specialAttributes[$key] = $number->attributes();
+            }
+        }
+    }
+    return $specialAttributes;
+}
+
+
+/**
+ * Restore special attributes in given target phone book
+ *
+ * @param   SimpleXMLElement    $XMLTargetPhoneBook
+ * @param   array               $oldSpecialAttributes [] or map key => attributes
+ * @return  SimpleXMLElement    out phonebook with restored special attributes
+ */
+function restoreSpecialAttributes(SimpleXMLElement $XMLTargetPhoneBook, array $oldSpecialAttributes)
+{
+    if (count($oldSpecialAttributes)==0) {
+        return $XMLTargetPhoneBook;
+    }
+    echo "Restoring old special attributes (quickdial, vanity)\n";
+    foreach ($XMLTargetPhoneBook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
+            // build unique key: {normalized-phone-number}@{vCard UUID} mapping to quick dial number
+            // normalizing means: remove all non-"+" and non-number characters like SPACE, MINUS, SLASH...
+            $key = preg_replace("/[^\+0-9]/", "", $number)."@".$contact->carddav_uid;
+            if (array_key_exists($key, $oldSpecialAttributes)) {
+                if (property_exists($oldSpecialAttributes[$key], "quickdial")) {
```

Comment:
quickdial and vanity are handled exactly the same- move code into function for loop:

````
foreach (['quickdial','vanity'] as $attribute) {
    ...
}
````

Reply:
Good. Did't know this trick in PHP.

Reply:
See #177 for an example